socks5.c

```diff
@@ -0,0 +1,337 @@
+/*
+
+  Copyright (c) 2018 Joseph deBlaquiere, Martin Sustrik
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files (the "Software"),
+  to deal in the Software without restriction, including without limitation
+  the rights to use, copy, modify, merge, publish, distribute, sublicense,
+  and/or sell copies of the Software, and to permit persons to whom
+  the Software is furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included
+  in all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+  IN THE SOFTWARE.
+
+*/
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define DILL_DISABLE_RAW_NAMES
+#include "libdillimpl.h"
+#include "utils.h"
+
+dill_unique_id(dill_socks5_type);
+
+static void *dill_socks5_hquery(struct dill_hvfs *hvfs, const void *type);
+static void dill_socks5_hclose(struct dill_hvfs *hvfs);
+static int dill_socks5_bsendl(struct dill_bsock_vfs *bvfs,
+    struct dill_iolist *first, struct dill_iolist *last, int64_t deadline);
+static int dill_socks5_brecvl(struct dill_bsock_vfs *bvfs,
+    struct dill_iolist *first, struct dill_iolist *last, int64_t deadline);
+
+// SOCKS5 as defined in RFC 1928
+
+struct dill_socks5_sock {
+    struct dill_hvfs hvfs;
+    struct dill_bsock_vfs bvfs;
+    /* Underlying SUFFIX socket. */
```

Comment:
Remove the comment. It's untrue. The underlying socket can be any bytestream socket.